### PR TITLE
WGXNP2

### PR DESCRIPTION
**Pull request by FeNToNelele**
Creates a VM with a service account. The service account and the compute disk is received from a separate module.

The old idea to create a new file on VM was to use a provisioner. However, it is a wrong and "last resort" solution, so I use a startup script instead.

### CODE REVIEW

File: WGXNP2/Beadando/Files/task [hunk 1]
```diff
@@ -0,0 +1 @@
+done
```
Comment (s1mark) [inline WGXNP2/Beadando/Files/task:1]: this is not necessary

File: WGXNP2/Beadando/Modules/ComputeDisk/provider.tf [hunk 2]
```diff
@@ -0,0 +1,4 @@
+provider "google" {
```
Comment (s1mark) [inline WGXNP2/Beadando/Modules/ComputeDisk/provider.tf:1]: the provider is only necessary where you invoke terraform from.
The modules dont need provider block

File: WGXNP2/Beadando/ServiceAccount/provider.tf [hunk 3]
```diff
@@ -0,0 +1,11 @@
+provider "google" {
```
Comment (s1mark) [inline WGXNP2/Beadando/ServiceAccount/provider.tf:1]: the provider is only necessary where you invoke terraform from.
The modules dont need provider block

File: WGXNP2/Beadando/ServiceAccount/provider.tf [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+    project = var.project_name
+    region = "europe-central2"
+}
+
+terraform {
```
Comment (s1mark) [inline WGXNP2/Beadando/ServiceAccount/provider.tf:6]: the backend configuration should be located where you invoke the terraform frome

File: WGXNP2/Beadando/VM/main.tf [hunk 5]
```diff
@@ -0,0 +1,34 @@
+module "service_account" {
```
Comment (s1mark) [inline WGXNP2/Beadando/VM/main.tf:1]: so it seems this the place where you orchestrate the whole codebase...
this means that the gcs backend and the provider should be located on this level

**Comment by s1mark** [comment]
After the fixes, try running terraform plan and post the output

**Comment by FeNToNelele** [comment]
> After the fixes, try running terraform plan and post the output

It should work, thank you for the suggestions.

PS C:\Users\Gabor\Desktop\Cloud\cloud\WGXNP2\Beadando\VM> terraform plan    

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following   
symbols:
  + create
 <= read (data resources)

Terraform will perform the following actions:

  # data.google_compute_instance.vm_data will be read during apply
  # (depends on a resource or a module with changes pending)
 <= data "google_compute_instance" "vm_data" {
      + advanced_machine_features    = (known after apply)
      + allow_stopping_for_update    = (known after apply)
      + attached_disk                = (known after apply)
      + boot_disk                    = (known after apply)
      + can_ip_forward               = (known after apply)
      + confidential_instance_config = (known after apply)
      + cpu_platform                 = (known after apply)
      + current_status               = (known after apply)
      + deletion_protection          = (known after apply)
      + description                  = (known after apply)
      + desired_status               = (known after apply)
      + enable_display               = (known after apply)
      + guest_accelerator            = (known after apply)
      + hostname                     = (known after apply)
      + id                           = (known after apply)
      + instance_id                  = (known after apply)
      + label_fingerprint            = (known after apply)
      + labels                       = (known after apply)
      + machine_type                 = (known after apply)
      + metadata                     = (known after apply)
      + metadata_fingerprint         = (known after apply)
      + metadata_startup_script      = (known after apply)
      + min_cpu_platform             = (known after apply)
      + name                         = "wgxnp2"
      + network_interface            = (known after apply)
      + reservation_affinity         = (known after apply)
      + resource_policies            = (known after apply)
      + scheduling                   = (known after apply)
      + scratch_disk                 = (known after apply)
      + service_account              = (known after apply)
      + shielded_instance_config     = (known after apply)
      + tags                         = (known after apply)
      + tags_fingerprint             = (known after apply)
    }

  # google_compute_instance.default will be created
  + resource "google_compute_instance" "default" {
      + can_ip_forward          = false
      + cpu_platform            = (known after apply)
      + current_status          = (known after apply)
      + deletion_protection     = false
      + guest_accelerator       = (known after apply)
      + id                      = (known after apply)
      + instance_id             = (known after apply)
      + label_fingerprint       = (known after apply)
      + machine_type            = "f1-micro"
      + metadata_fingerprint    = (known after apply)
      + metadata_startup_script = "echo done > ~/task"
      + min_cpu_platform        = (known after apply)
      + name                    = "wgxnp2"
      + project                 = (known after apply)
      + self_link               = (known after apply)
      + tags_fingerprint        = (known after apply)
      + zone                    = "europe-central2"

      + boot_disk {
          + auto_delete                = true
          + device_name                = (known after apply)
          + disk_encryption_key_sha256 = (known after apply)
          + kms_key_self_link          = (known after apply)
          + mode                       = "READ_WRITE"
          + source                     = (known after apply)

          + initialize_params {
              + image  = "debian-cloud/debian-11"
              + labels = (known after apply)
              + size   = (known after apply)
              + type   = (known after apply)
            }
        }

      + confidential_instance_config {
          + enable_confidential_compute = (known after apply)
        }

      + network_interface {
          + ipv6_access_type   = (known after apply)
          + name               = (known after apply)
          + network            = "default"
          + network_ip         = (known after apply)
          + stack_type         = (known after apply)
          + subnetwork         = (known after apply)
          + subnetwork_project = (known after apply)
        }

      + reservation_affinity {
          + type = (known after apply)

          + specific_reservation {
              + key    = (known after apply)
              + values = (known after apply)
            }
        }

      + scheduling {
          + automatic_restart           = (known after apply)
          + instance_termination_action = (known after apply)
          + min_node_cpus               = (known after apply)
          + on_host_maintenance         = (known after apply)
          + preemptible                 = (known after apply)
          + provisioning_model          = (known after apply)

          + node_affinities {
              + key      = (known after apply)
              + operator = (known after apply)
              + values   = (known after apply)
            }
        }

      + service_account {
          + email  = (known after apply)
          + scopes = [
              + "https://www.googleapis.com/auth/monitoring",
            ]
        }
    }

  # module.compute_disk.google_compute_disk.default will be created
  + resource "google_compute_disk" "default" {
      + creation_timestamp        = (known after apply)
      + id                        = (known after apply)
      + label_fingerprint         = (known after apply)
      + last_attach_timestamp     = (known after apply)
      + last_detach_timestamp     = (known after apply)
      + name                      = "wgxnp2"
      + physical_block_size_bytes = 4096
      + project                   = (known after apply)
      + provisioned_iops          = (known after apply)
      + self_link                 = (known after apply)
      + size                      = 10
      + source_disk_id            = (known after apply)
      + source_image_id           = (known after apply)
      + source_snapshot_id        = (known after apply)
      + type                      = "pd-ssd"
      + users                     = (known after apply)
      + zone                      = "europe-central2"
    }

  # module.service_account.google_service_account.service_account will be created
  + resource "google_service_account" "service_account" {
      + account_id   = "wgxnp2"
      + disabled     = false
      + display_name = "WGXNP2"
      + email        = (known after apply)
      + id           = (known after apply)
      + member       = (known after apply)
      + name         = (known after apply)
      + project      = (known after apply)
      + unique_id    = (known after apply)
    }

Plan: 3 to add, 0 to change, 0 to destroy.

Changes to Outputs:
  + creation_year = "2022"
  + creator       = "Gabor Sarosi"
  + vm_name       = "wgxnp2"

**Comment by s1mark** [comment]
Yeah the plan look ok. The real test is when you apply it.

**Comment by s1mark** [comment]
Thanks for fixing what we discussed. Did the apply go through without issues?

**Pull request merged**